tools/module_diag_hailcast.F90

```diff
@@ -0,0 +1,1870 @@
+      CLOUDON=1  !we'll eventually turn cloud "off" once updraft past time limit
+
+      !Start time loop.
+      DO WHILE (sec .lt. TAU)
+         sec = sec + secdel
+
+         !!!!!!!!!!!!!!!!!!  5. CALCULATE PARAMETERS  !!!!!!!!!!!!!!!!!
+         !!!              CALCULATE UPDRAFT PROPERTIES              !!!
+         !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+         !Intepolate vertical velocity to our new pressure level
+         CALL INTERP(VUU_pert,VUMAX,P,IFOUT,PA,nz)
+         !print *, 'INTERP VU: ', VU, P
+
+         !Outside pressure levels?  If so, exit loop
+         IF (IFOUT.EQ.1) GOTO 100
+
+         !Sine wave multiplier on updraft strength
+         IF (SEC .GT. 0.0 .AND. SEC .LT. RTIME) THEN
+            VUCORE = VUMAX * (0.5*SIN((3.14159*SEC)/(RTIME))+0.5)*1.2
+            VU = VUCORE
+         ELSEIF (SEC .GE. RTIME) THEN
+            VU = 0.0
+            CLOUDON = 0
+         ENDIF
+
+         !Calculate terminal velocity of the hailstone
+         ! (use previous values)
+         CALL TERMINL(DENSA,DENSE,D,VT,TC)
+
+         !Actual velocity of hailstone (upwards positive)
+         V = VU - VT
+
+         !Use hydrostatic eq'n to calc height of next level
+         P = P - DENSA*g*V*secdel
+         Z = Z + V*secdel
+
+         !Interpolate cloud temp, qvapor at new p-level
+         CALL INTERP(TCA,TC,P,IFOUT,PA,nz)
+         CALL INTERP(RA,RS,P,IFOUT,PA,nz)
+
+         !New density of in-cloud air
+         DENSA=P/(r_d*(1.+0.609*RS/(1.+RS))*TC)
+
+         !Interpolate liquid, frozen water mix ratio at new level
+         CALL INTERP(RIA,RI,P,IFOUT,PA,nz)
+         CALL INTERP(RWA_new,RW,P,IFOUT,PA,nz)
+         XI = RI * DENSA * CLOUDON
+         XW = RW * DENSA * CLOUDON
+         IF( (XW+XI).GT.0) THEN
+           PC = XI / (XW+XI)
+         ELSE
+           PC = 1.
+         ENDIF
+
+
+        ! SATURATION VAPOUR DENSITY DIFFERENCE BETWTEEN STONE AND CLOUD
+        CALL VAPORCLOSE(DELRW,PC,TS,TC,ITYPE)
+
+
+        !!!!!!!!!!!!!!!!!!  6. STONE'S MASS GROWTH !!!!!!!!!!!!!!!!!!!!
+        CALL MASSAGR(D,GM,GM1,GMW,GMI,DGM,DGMW,DGMI,DGMV,DI,ANU,RE,AE,&
+                 TC,TS,P,DENSE,DENSA,FW,VT,XW,XI,secdel,ITYPE,DELRW)
+
+
+        !!!!!!!!!!!!!!!!!!  7. HEAT BUDGET OF HAILSTONE !!!!!!!!!!!!!!!
+        CALL HEATBUD(TS,TSm1,TSm2,FW,TC,VT,DELRW,D,DENSA,GM1,GM,DGM,DGMW,  &
+                     DGMV,DGMI,GMW,GMI,DI,ANU,RE,AE,secdel,ITYPE,P)
+
+
+        !!!!! 8. TEST DIAMETER OF STONE AND HEIGHT ABOVE GROUND !!!!!!!
+        !!!  TEST IF DIAMETER OF STONE IS GREATER THAN CRITICAL LIMIT, IF SO
+        !!!  BREAK UP
+        WATER=FW*GM  !KG
+        ! CRTICAL MASS CAPABLE OF BEING "SUPPORTED" ON THE STONE'S SURFACE
+        CRIT = 2.0E-4
+        IF (WATER.GT.CRIT)THEN
+           CALL BREAKUP(DENSE,D,GM,FW,CRIT)
+        ENDIF
+
+        !!! Has stone reached below cloud base?
+        !IF (Z .LE. 0) GOTO 200
+        IF (Z .LE. ZBAS) then
+           GOTO 200
```

Comment:
Please rewrite the code here to remove all the GOTO statement, it is not allowed in operational model code.